### DISCUSSION TITLE
Is there a way to get git blame for a specific line in helix?

### DISCUSSION THREAD

**Discussion by sanket143** [helix-editor/helix#6421] [Q&A] [answered]
In VS code, I used to have git lens for git blame for a specific line, is there a way to get git blame for a specific line in the current implementation of helix? or a convenient workaround?

**Comment by danillos** [comment, marked as answer]
Currently, it is not possible.

But a workaround could be possible using this PR https://github.com/helix-editor/helix/pull/3393 when it includes a variable like `%val{linenumber}`.

We could build a script with a keybinding to show the result of the script in a popup like:

```
F7 = [':sh git_blame.sh %val{filename}:%val{linenumber}']
```

**Reply by Septias** [reply]
where is git_blame.sh defined?

**Reply by jm355** [reply]
`git log -n 1 --format='format:%h %an %s' --no-patch -L<line #>:<filename>`

**Reply by Septias** [reply]
Ty

**Comment by rcorre** [comment]
The PR @danillos referenced was closed, and https://github.com/helix-editor/helix/pull/12527 merged instead. This allows a similar solution but with slightly different syntax. There's a few suggestions starting at https://github.com/helix-editor/helix/issues/3035#issuecomment-2715731119. Combining the ideas of @danillos, @DoctorRyner, and @rockboynton, this is my current binding:

```toml
B = ":sh git log -n 5 --format='format:%%h (%%an: %%ar) %%s' --no-patch -L%{cursor_line},+1:%{buffer_name}"
```

Note that `%` now needs to be escaped as `%%`, and the `-L` flag requires both a start and end line number (I'm using the relative offset `+1`). Compared to `blame`, `log` has the advantage of working even if you use it on a dirty line. Since the `sh` output can show multiple lines, I'm using `-5` to see the last 5 changes.

**Comment by thomasaarholt** [comment]
I wrote (with claude) a bash script that extracts the author, time, message and commit url (it gets the PR url instead if the PR number is present in the commit message). By default the script produces the information below, but if you pass `--url-only`, then it only returns the url, so you can pass it to the browser through the CLI. Enjoy, and forks are more than welcome.

![image](https://github.com/user-attachments/assets/c009fbc4-e4e5-4d3e-a817-a9881c7b1aab)


Add [this `blame` bash script](https://gist.github.com/thomasaarholt/2659407306e83edde4262a398aaf9f43) somewhere on your path (or use the full path in your helix config).

Then pass the following to your helix config:
```toml
[keys.normal.space]
B = ":sh blame %{buffer_name} %{cursor_line}" # Show blame information
U = ":sh blame --url-only %{cursor_line} %{buffer_name} | xargs -I{} xdg-open {}" # Open PR or commit url in browser
```

**Reply by a3lem** [reply]
This is excellent! Thanks for sharing!

**Reply by fidelicura** [reply]
Perfect, just perfect! Thanks!

**Comment by nik-rev** [comment]
There is now an easier way if you build https://github.com/helix-editor/helix/pull/13133

Use `space + B` and you will get the blame for the current line

**Comment by pxp9** [comment]
Hey a little update on this thread, 

This PR allows the commands expansions which was merged on February. #12527

This PR is working in the proper feature as @nik-rev mentioned  #13133 

There is no current stable release which have one of these features that allow you to blame properly your teammate.

Last release at the time I am writing this is 25.01.1 which was released on January.

**Reply by djpowers** [reply]
Just updating that as of July 15, 2025 the expansions feature is available: https://helix-editor.com/news/release-25-07-highlights/#expansions

This allows printing git blame information to the statusline as in this example: https://docs.helix-editor.com/command-line.html#expansions
